File: server/camcops_server/tasks/psychiatricclerking.py
"""
camcops_server/tasks/psychiatricclerking.py

===============================================================================

    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

"""

from typing import Dict, List, Optional

import cardinal_pythonlib.rnc_web as ws
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql.sqltypes import UnicodeText

from camcops_server.cc_modules.cc_constants import CssClass
from camcops_server.cc_modules.cc_ctvinfo import CtvInfo
from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_snomed import (
    SnomedConcept,
    SnomedExpression,
    SnomedLookup,
)
from camcops_server.cc_modules.cc_sqlalchemy import Base
from camcops_server.cc_modules.cc_task import (
    Task,
    TaskHasClinicianMixin,
    TaskHasPatientMixin,
)


# =============================================================================
# PsychiatricClerking
# =============================================================================


class PsychiatricClerking(  # type: ignore[misc]
    TaskHasPatientMixin, TaskHasClinicianMixin, Task, Base
):
    """
    Server implementation of the Clerking task.
    """

    __tablename__ = "psychiatricclerking"
    shortname = "Clerking"
    info_filename_stem = "clinical"

    # FIELDSPEC_A = CLINICIAN_FIELDSPECS  # replaced by has_clinician, then by TaskHasClinicianMixin  # noqa

    location: Mapped[Optional[str]] = mapped_column("location", UnicodeText)
    contact_type: Mapped[Optional[str]] = mapped_column(UnicodeText)
    reason_for_contact: Mapped[Optional[str]] = mapped_column(UnicodeText)
    presenting_issue: Mapped[Optional[str]] = mapped_column(UnicodeText)
    systems_review: Mapped[Optional[str]] = mapped_column(UnicodeText)
    collateral_history: Mapped[Optional[str]] = mapped_column(UnicodeText)

    diagnoses_psychiatric: Mapped[Optional[str]] = mapped_column(UnicodeText)
    diagnoses_medical: Mapped[Optional[str]] = mapped_column(UnicodeText)
    operations_procedures: Mapped[Optional[str]] = mapped_column(UnicodeText)
    allergies_adverse_reactions: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    medications: Mapped[Optional[str]] = mapped_column(UnicodeText)
    recreational_drug_use: Mapped[Optional[str]] = mapped_column(UnicodeText)
    family_history: Mapped[Optional[str]] = mapped_column(UnicodeText)
    developmental_history: Mapped[Optional[str]] = mapped_column(UnicodeText)
    personal_history: Mapped[Optional[str]] = mapped_column(UnicodeText)
    premorbid_personality: Mapped[Optional[str]] = mapped_column(UnicodeText)
    forensic_history: Mapped[Optional[str]] = mapped_column(UnicodeText)
    current_social_situation: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )

    mse_appearance_behaviour: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    mse_speech: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_mood_subjective: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_mood_objective: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_thought_form: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_thought_content: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_perception: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_cognition: Mapped[Optional[str]] = mapped_column(UnicodeText)
    mse_insight: Mapped[Optional[str]] = mapped_column(UnicodeText)

    physical_examination_general: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    physical_examination_cardiovascular: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    physical_examination_respiratory: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    physical_examination_abdominal: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )
    physical_examination_neurological: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )

    assessment_scales: Mapped[Optional[str]] = mapped_column(UnicodeText)
    investigations_results: Mapped[Optional[str]] = mapped_column(UnicodeText)

    safety_alerts: Mapped[Optional[str]] = mapped_column(UnicodeText)
    risk_assessment: Mapped[Optional[str]] = mapped_column(UnicodeText)
    relevant_legal_information: Mapped[Optional[str]] = mapped_column(
        UnicodeText
    )

    current_problems: Mapped[Optional[str]] = mapped_column(UnicodeText)
    patient_carer_concerns: Mapped[Optional[str]] = mapped_column(UnicodeText)
    impression: Mapped[Optional[str]] = mapped_column(UnicodeText)
    management_plan: Mapped[Optional[str]] = mapped_column(UnicodeText)
    information_given: Mapped[Optional[str]] = mapped_column(UnicodeText)

    FIELDS_B = [
        "location",
        "contact_type",
        "reason_for_contact",
        "presenting_issue",
        "systems_review",
        "collateral_history",
    ]
    FIELDS_C = [
        "diagnoses_psychiatric",
        "diagnoses_medical",
        "operations_procedures",
        "allergies_adverse_reactions",
        "medications",
        "recreational_drug_use",
        "family_history",
        "developmental_history",
        "personal_history",
        "premorbid_personality",
        "forensic_history",
        "current_social_situation",
    ]
    FIELDS_MSE = [
        "mse_appearance_behaviour",
        "mse_speech",
        "mse_mood_subjective",
        "mse_mood_objective",
        "mse_thought_form",
        "mse_thought_content",
        "mse_perception",
        "mse_cognition",
        "mse_insight",
    ]
    FIELDS_PE = [
        "physical_examination_general",
        "physical_examination_cardiovascular",
        "physical_examination_respiratory",
        "physical_examination_abdominal",
        "physical_examination_neurological",
    ]
    FIELDS_D = ["assessment_scales", "investigations_results"]
    FIELDS_E = [
        "safety_alerts",
        "risk_assessment",
        "relevant_legal_information",
    ]
    FIELDS_F = [
        "current_problems",
        "patient_carer_concerns",
        "impression",
        "management_plan",
        "information_given",
    ]

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Psychiatric clerking")

    def get_ctv_heading(
        self, req: CamcopsRequest, wstringname: str
    ) -> CtvInfo:
        return CtvInfo(
            heading=self.wxstring(req, wstringname), skip_if_no_content=False
        )

    def get_ctv_subheading(
        self, req: CamcopsRequest, wstringname: str
    ) -> CtvInfo:
        return CtvInfo(
            subheading=self.wxstring(req, wstringname),
            skip_if_no_content=False,
        )

    def get_ctv_description_content(
        self, req: CamcopsRequest, x: str
    ) -> CtvInfo:
        return CtvInfo(
            description=self.wxstring(req, x),
            content=ws.webify(getattr(self, x)),
        )

    def get_clinical_text(self, req: CamcopsRequest) -> List[CtvInfo]:
        infolist = [self.get_ctv_heading(req, "heading_current_contact")]
        for x in self.FIELDS_B:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(self.get_ctv_heading(req, "heading_background"))
        for x in self.FIELDS_C:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(
            self.get_ctv_heading(req, "heading_examination_investigations")
        )
        infolist.append(
            self.get_ctv_subheading(req, "mental_state_examination")
        )
        for x in self.FIELDS_MSE:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(self.get_ctv_subheading(req, "physical_examination"))
        for x in self.FIELDS_PE:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(
            self.get_ctv_subheading(req, "assessments_and_investigations")
        )
        for x in self.FIELDS_D:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(self.get_ctv_heading(req, "heading_risk_legal"))
        for x in self.FIELDS_E:
            infolist.append(self.get_ctv_description_content(req, x))
        infolist.append(self.get_ctv_heading(req, "heading_summary_plan"))
        for x in self.FIELDS_F:
            infolist.append(self.get_ctv_description_content(req, x))
        return infolist

    # noinspection PyMethodOverriding
    @staticmethod
    def is_complete() -> bool:
        return True

    def heading(self, req: CamcopsRequest, wstringname: str) -> str:
        return '<div class="{CssClass.HEADING}">{s}</div>'.format(
            CssClass=CssClass, s=self.wxstring(req, wstringname)
        )

    def subheading(self, req: CamcopsRequest, wstringname: str) -> str:
        return '<div class="{CssClass.SUBHEADING}">{s}</div>'.format(
            CssClass=CssClass, s=self.wxstring(req, wstringname)
        )

    def subsubheading(self, req: CamcopsRequest, wstringname: str) -> str:
        return '<div class="{CssClass.SUBSUBHEADING}">{s}</div>'.format(
            CssClass=CssClass, s=self.wxstring(req, wstringname)
        )

    def subhead_text(self, req: CamcopsRequest, fieldname: str) -> str:
        return self.subheading(req, fieldname) + "<div><b>{}</b></div>".format(
            ws.webify(getattr(self, fieldname))
        )

    def subsubhead_text(self, req: CamcopsRequest, fieldname: str) -> str:
        return (
            self.subsubheading(req, fieldname)
            + f"<div><b>{ws.webify(getattr(self, fieldname))}</b></div>"
        )

    def get_task_html(self, req: CamcopsRequest) -> str:
        # Avoid tables - PDF generator crashes if text is too long.
        html = ""
        html += self.heading(req, "heading_current_contact")
        for x in self.FIELDS_B:
            html += self.subhead_text(req, x)
        html += self.heading(req, "heading_background")
        for x in self.FIELDS_C:
            html += self.subhead_text(req, x)
        html += self.heading(req, "heading_examination_investigations")
        html += self.subheading(req, "mental_state_examination")
        for x in self.FIELDS_MSE:
            html += self.subsubhead_text(req, x)
        html += self.subheading(req, "physical_examination")
        for x in self.FIELDS_PE:
            html += self.subsubhead_text(req, x)
        for x in self.FIELDS_D:
            html += self.subhead_text(req, x)
        html += self.heading(req, "heading_risk_legal")
        for x in self.FIELDS_E:
            html += self.subhead_text(req, x)
        html += self.heading(req, "heading_summary_plan")
        for x in self.FIELDS_F:
            html += self.subhead_text(req, x)
        return html

    def get_snomed_codes(self, req: CamcopsRequest) -> List[SnomedExpression]:
        refinement = {}  # type: Dict[SnomedConcept, str]

        def add(snomed_lookup: str, contents: Optional[str]) -> None:
            if not contents:
                return
            nonlocal refinement
            concept = req.snomed(snomed_lookup)
            refinement[concept] = contents

        # not location
        # not contact type
        add(SnomedLookup.PSYCLERK_REASON_FOR_REFERRAL, self.reason_for_contact)
        add(SnomedLookup.PSYCLERK_PRESENTING_ISSUE, self.presenting_issue)
        add(SnomedLookup.PSYCLERK_SYSTEMS_REVIEW, self.systems_review)
        add(SnomedLookup.PSYCLERK_COLLATERAL_HISTORY, self.collateral_history)

        add(
            SnomedLookup.PSYCLERK_PAST_MEDICAL_SURGICAL_MENTAL_HEALTH_HISTORY,
            self.diagnoses_medical,
        )
        add(
            SnomedLookup.PSYCLERK_PAST_MEDICAL_SURGICAL_MENTAL_HEALTH_HISTORY,
            self.diagnoses_psychiatric,
        )
        add(SnomedLookup.PSYCLERK_PROCEDURES, self.operations_procedures)
        add(
            SnomedLookup.PSYCLERK_ALLERGIES_ADVERSE_REACTIONS,
            self.allergies_adverse_reactions,
        )
        add(
            SnomedLookup.PSYCLERK_MEDICATIONS_MEDICAL_DEVICES, self.medications
        )
        add(
            SnomedLookup.PSYCLERK_DRUG_SUBSTANCE_USE,
            self.recreational_drug_use,
        )
        add(SnomedLookup.PSYCLERK_FAMILY_HISTORY, self.family_history)
        add(
            SnomedLookup.PSYCLERK_DEVELOPMENTAL_HISTORY,
            self.developmental_history,
        )
        add(
            SnomedLookup.PSYCLERK_SOCIAL_PERSONAL_HISTORY,
            self.personal_history,
        )
        add(SnomedLookup.PSYCLERK_PERSONALITY, self.premorbid_personality)
        add(
            SnomedLookup.PSYCLERK_PRISON_RECORD_CRIMINAL_ACTIVITY,
            self.forensic_history,
        )
        add(
            SnomedLookup.PSYCLERK_SOCIAL_HISTORY_BASELINE,
            self.current_social_situation,
        )

        add(
            SnomedLookup.PSYCLERK_MSE_APPEARANCE, self.mse_appearance_behaviour
        )  # duplication
        add(
            SnomedLookup.PSYCLERK_MSE_BEHAVIOUR, self.mse_appearance_behaviour
        )  # duplication
        add(SnomedLookup.PSYCLERK_MSE_MOOD, self.mse_mood_subjective)  # close
        add(SnomedLookup.PSYCLERK_MSE_AFFECT, self.mse_mood_objective)
        # ... Logic here: "objective mood" is certainly affect (emotional
        # weather). "Subjective mood" is both mood (emotional climate) and
        # affect. Not perfect, but reasonable.
        add(SnomedLookup.PSYCLERK_MSE_THOUGHT, self.mse_thought_form)
        add(SnomedLookup.PSYCLERK_MSE_THOUGHT, self.mse_thought_content)
        # ... No way of disambiguating the two in SNOMED-CT.
        add(SnomedLookup.PSYCLERK_MSE_PERCEPTION, self.mse_perception)
        add(SnomedLookup.PSYCLERK_MSE_COGNITION, self.mse_cognition)
        add(SnomedLookup.PSYCLERK_MSE_INSIGHT, self.mse_insight)

        add(
            SnomedLookup.PSYCLERK_PHYSEXAM_GENERAL,
            self.physical_examination_general,
        )
        add(
            SnomedLookup.PSYCLERK_PHYSEXAM_CARDIOVASCULAR,
            self.physical_examination_cardiovascular,
        )
        add(
            SnomedLookup.PSYCLERK_PHYSEXAM_RESPIRATORY,
            self.physical_examination_respiratory,
        )
        add(
            SnomedLookup.PSYCLERK_PHYSEXAM_ABDOMINAL,
            self.physical_examination_abdominal,
        )
        add(
            SnomedLookup.PSYCLERK_PHYSEXAM_NEUROLOGICAL,
            self.physical_examination_neurological,
        )

        add(SnomedLookup.PSYCLERK_ASSESSMENT_SCALES, self.assessment_scales)
        add(
            SnomedLookup.PSYCLERK_INVESTIGATIONS_RESULTS,
            self.investigations_results,
        )

        add(SnomedLookup.PSYCLERK_SAFETY_ALERTS, self.safety_alerts)
        add(SnomedLookup.PSYCLERK_RISK_ASSESSMENT, self.risk_assessment)
        add(
            SnomedLookup.PSYCLERK_RELEVANT_LEGAL_INFORMATION,
            self.relevant_legal_information,
        )

        add(SnomedLookup.PSYCLERK_CURRENT_PROBLEMS, self.current_problems)
        add(
            SnomedLookup.PSYCLERK_PATIENT_CARER_CONCERNS,
            self.patient_carer_concerns,
        )
        add(SnomedLookup.PSYCLERK_CLINICAL_NARRATIVE, self.impression)
        add(SnomedLookup.PSYCLERK_MANAGEMENT_PLAN, self.management_plan)
        add(SnomedLookup.PSYCLERK_INFORMATION_GIVEN, self.information_given)

        codes = [
            SnomedExpression(
                req.snomed(
                    SnomedLookup.DIAGNOSTIC_PSYCHIATRIC_INTERVIEW_PROCEDURE
                ),
                refinement=refinement or None,  # type: ignore[arg-type]
            )
        ]
        return codes

File: tablet_qt/tasks/psychiatricclerking.cpp
/*
    Copyright (C) 2012-2018 Rudolf Cardinal (rudolf@pobox.com).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <http://www.gnu.org/licenses/>.
*/

#include "psychiatricclerking.h"
#include "common/textconst.h"
#include "lib/stringfunc.h"
#include "questionnairelib/questionnaire.h"
#include "questionnairelib/quheading.h"
#include "questionnairelib/qulineedit.h"
#include "questionnairelib/qutext.h"
#include "questionnairelib/qutextedit.h"
#include "tasklib/taskfactory.h"

const QString PsychiatricClerking::PSYCLERK_TABLENAME("psychiatricclerking");

const QString LOCATION("location");
const QString CONTACT_TYPE("contact_type");
const QString REASON_FOR_CONTACT("reason_for_contact");
const QString PRESENTING_ISSUE("presenting_issue");
const QString SYSTEMS_REVIEW("systems_review");
const QString COLLATERAL_HISTORY("collateral_history");

const QString DIAGNOSES_PSYCHIATRIC("diagnoses_psychiatric");
const QString DIAGNOSES_MEDICAL("diagnoses_medical");
const QString OPERATIONS_PROCEDURES("operations_procedures");
const QString ALLERGIES_ADVERSE_REACTIONS("allergies_adverse_reactions");
const QString MEDICATIONS("medications");
const QString RECREATIONAL_DRUG_USE("recreational_drug_use");
const QString FAMILY_HISTORY("family_history");
const QString DEVELOPMENTAL_HISTORY("developmental_history");
const QString PERSONAL_HISTORY("personal_history");
const QString PREMORBID_PERSONALITY("premorbid_personality");
const QString FORENSIC_HISTORY("forensic_history");
const QString CURRENT_SOCIAL_SITUATION("current_social_situation");

const QString MSE_APPEARANCE_BEHAVIOUR("mse_appearance_behaviour");
const QString MSE_SPEECH("mse_speech");
const QString MSE_MOOD_SUBJECTIVE("mse_mood_subjective");
const QString MSE_MOOD_OBJECTIVE("mse_mood_objective");
const QString MSE_THOUGHT_FORM("mse_thought_form");
const QString MSE_THOUGHT_CONTENT("mse_thought_content");
const QString MSE_PERCEPTION("mse_perception");
const QString MSE_COGNITION("mse_cognition");
const QString MSE_INSIGHT("mse_insight");

const QString PHYSICAL_EXAMINATION_GENERAL("physical_examination_general");
const QString PHYSICAL_EXAMINATION_CARDIOVASCULAR("physical_examination_cardiovascular");
const QString PHYSICAL_EXAMINATION_RESPIRATORY("physical_examination_respiratory");
const QString PHYSICAL_EXAMINATION_ABDOMINAL("physical_examination_abdominal");
const QString PHYSICAL_EXAMINATION_NEUROLOGICAL("physical_examination_neurological");

const QString ASSESSMENT_SCALES("assessment_scales");
const QString INVESTIGATIONS_RESULTS("investigations_results");

const QString SAFETY_ALERTS("safety_alerts");
const QString RISK_ASSESSMENT("risk_assessment");
const QString RELEVANT_LEGAL_INFORMATION("relevant_legal_information");

const QString CURRENT_PROBLEMS("current_problems");
const QString PATIENT_CARER_CONCERNS("patient_carer_concerns");
const QString IMPRESSION("impression");
const QString MANAGEMENT_PLAN("management_plan");
const QString INFORMATION_GIVEN("information_given");

const QStringList EXTRAFIELDS_B{
    LOCATION,
    CONTACT_TYPE,
    REASON_FOR_CONTACT,
    PRESENTING_ISSUE,
    SYSTEMS_REVIEW,
    COLLATERAL_HISTORY,
};
const QStringList EXTRAFIELDS_C{
    DIAGNOSES_PSYCHIATRIC,
    DIAGNOSES_MEDICAL,
    OPERATIONS_PROCEDURES,
    ALLERGIES_ADVERSE_REACTIONS,
    MEDICATIONS,
    RECREATIONAL_DRUG_USE,
    FAMILY_HISTORY,
    DEVELOPMENTAL_HISTORY,
    PERSONAL_HISTORY,
    PREMORBID_PERSONALITY,
    FORENSIC_HISTORY,
    CURRENT_SOCIAL_SITUATION,
};
const QStringList EXTRAFIELDS_MSE{
    MSE_APPEARANCE_BEHAVIOUR,
    MSE_SPEECH,
    MSE_MOOD_SUBJECTIVE,
    MSE_MOOD_OBJECTIVE,
    MSE_THOUGHT_FORM,
    MSE_THOUGHT_CONTENT,
    MSE_PERCEPTION,
    MSE_COGNITION,
    MSE_INSIGHT,
};
const QStringList EXTRAFIELDS_PE{
    PHYSICAL_EXAMINATION_GENERAL,
    PHYSICAL_EXAMINATION_CARDIOVASCULAR,
    PHYSICAL_EXAMINATION_RESPIRATORY,
    PHYSICAL_EXAMINATION_ABDOMINAL,
    PHYSICAL_EXAMINATION_NEUROLOGICAL,
};
const QStringList EXTRAFIELDS_D{
    ASSESSMENT_SCALES,
    INVESTIGATIONS_RESULTS,
};
const QStringList EXTRAFIELDS_E{
    SAFETY_ALERTS,
    RISK_ASSESSMENT,
    RELEVANT_LEGAL_INFORMATION,
};
const QStringList EXTRAFIELDS_F{
    CURRENT_PROBLEMS,
    PATIENT_CARER_CONCERNS,
    IMPRESSION,
    MANAGEMENT_PLAN,
    INFORMATION_GIVEN,
};


void initializePsychiatricClerking(TaskFactory& factory)
{
    static TaskRegistrar<PsychiatricClerking> registered(factory);
}


PsychiatricClerking::PsychiatricClerking(CamcopsApp& app, DatabaseManager& db,
                                         const int load_pk) :
    Task(app, db, PSYCLERK_TABLENAME, false, true, false)  // ... anon, clin, resp
{
    // In the Javascript version, EXTRAFIELDS_A was the set of clinician fields.
    addFields(EXTRAFIELDS_B, QVariant::String);
    addFields(EXTRAFIELDS_C, QVariant::String);
    addFields(EXTRAFIELDS_MSE, QVariant::String);
    addFields(EXTRAFIELDS_PE, QVariant::String);
    addFields(EXTRAFIELDS_D, QVariant::String);
    addFields(EXTRAFIELDS_E, QVariant::String);
    addFields(EXTRAFIELDS_F, QVariant::String);

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}


// ============================================================================
// Class info
// ============================================================================

QString PsychiatricClerking::shortname() const
{
    return "Clerking";
}


QString PsychiatricClerking::longname() const
{
    return tr("Clerking");
}


QString PsychiatricClerking::menusubtitle() const
{
    return tr("Standard psychiatric clerking form.");
}


QString PsychiatricClerking::infoFilenameStem() const
{
    return "clinical";
}


// ============================================================================
// Instance info
// ============================================================================

bool PsychiatricClerking::isComplete() const
{
    return true;
}


QStringList PsychiatricClerking::summary() const
{
    return QStringList{
        fieldSummary(LOCATION, textconst::LOCATION, ": "),
    };
}


QStringList PsychiatricClerking::detail() const
{
    QStringList lines = completenessInfo() + clinicianDetails();

    auto add = [this, &lines](const QStringList& fields) -> void {
        for (auto field : fields) {
            lines.append(fieldSummary(field, xstring(field), ": "));
        }
    };

    add(EXTRAFIELDS_B);
    add(EXTRAFIELDS_C);
    add(EXTRAFIELDS_MSE);
    add(EXTRAFIELDS_PE);
    add(EXTRAFIELDS_D);
    add(EXTRAFIELDS_E);
    add(EXTRAFIELDS_F);
    return lines;
}


OpenableWidget* PsychiatricClerking::editor(const bool read_only)
{
    QVector<QuElement*> elements{getClinicianQuestionnaireBlockRawPointer()};

    auto addGroup = [this, &elements](const QStringList& fields) -> void {
        for (auto field : fields) {
            elements.append(new QuText(xstring(field)));
            QuTextEdit* edit = new QuTextEdit(fieldRef(field, false));
            edit->setHint("");
            elements.append(edit);
        }
    };
    auto addHeading = [this, &elements](const QString& xstringname) -> void {
        elements.append(new QuHeading(xstring(xstringname)));
    };
    auto addSubheading = [this, &elements](const QString& xstringname) -> void {
        elements.append((new QuText(xstring(xstringname)))->setBold(true));
    };

    addHeading("heading_current_contact");
    addGroup(EXTRAFIELDS_B);

    addHeading("heading_background");
    addGroup(EXTRAFIELDS_C);

    addHeading("heading_examination_investigations");
    addSubheading("mental_state_examination");
    addGroup(EXTRAFIELDS_MSE);
    addSubheading("physical_examination");
    addGroup(EXTRAFIELDS_PE);
    addSubheading("assessments_and_investigations");
    addGroup(EXTRAFIELDS_D);

    addHeading("heading_risk_legal");
    addGroup(EXTRAFIELDS_E);

    addHeading("heading_summary_plan");
    addGroup(EXTRAFIELDS_F);

    QuPagePtr page((new QuPage(elements))->setTitle(xstring("title")));

    Questionnaire* questionnaire = new Questionnaire(m_app, {page});
    questionnaire->setType(QuPage::PageType::Clinician);
    questionnaire->setReadOnly(read_only);
    return questionnaire;
}
